### PR TITLE
Short links

### LINKED ISSUE #6309: URL shortening
## Detailed Description
Backend-implemented URL shortening service for sharing links. User clicks on the share button in the frontend, frontend sends the assembled URL to the backend, backend stores the full URL in the database and creates an ID, the ID is returned to the frontend, and copied as URL to the clipboard. 
The short URLs could be in the form of `https://webknossos.org/links/62bc164cb824c569c505b610` or we plug into the wklink.org domain. 
The number of sharing links could be potentially be very high, but I don't think that is a problem in practice.

## Context
The sharing links contain lots of useful configurations, but are crazy long. They take up so much space in email that the rest of the text get dwarfed.

### PR DESCRIPTION
**Pull request by frcroth**
### URL of deployed dev instance (used for testing):
- https://featureshortlinks.webknossos.xyz

### Steps to test (Backend):
- Create a shortlink (POST /api/shortLinks)
- Read a shortlink (GET /api/shortLinks/byKey/:key)
- Check sensible errors when giving invalid parameters

### Steps to test (end-to-end)
- In Annotation sharing modal, copy short link
- In new tab, redirect should work

### Issues:
- fixes part of #6309 
- This is just allows for the creation of links, routing shortlinks and converting them is not implemented

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [x] Ready for review

### CODE REVIEW

File: app/controllers/ShortlinkController.scala
Comment (fm3) [reply app/controllers/ShortlinkController.scala:38]: On second thought, it should probably be getByShortLink, since that is how you called this particular field in the case class.

**Comment by normanrz** [comment]
> Another open question: should there be a mechanism to retreive old short links? Or is it fair to just create a new one each time the user clicks on copy?

I don't think retrieving a list is necessary.

**Comment by fm3** [comment]
> I don't think retrieving a list is necessary.

Ok cool. @frcroth in this case, please remove the get-by-id route and its corresponding implementation in the DAO. If the frontend is not going to use it, I’d rather not have it dangling there (the db ids might even be guessable). In the database schema, though, I still think it makes (some) sense to have the id.

**Comment by frcroth** [comment]
> > I don't think retrieving a list is necessary.
> 
> Ok cool. @frcroth in this case, please remove the get-by-id route and its corresponding implementation in the DAO. If the frontend is not going to use it, I’d rather not have it dangling there (the db ids might even be guessable). In the database schema, though, I still think it makes (some) sense to have the id.

The DAO method has to remain because it is used to verify the creation, but the controller method and the route can be removed. Should the other route still be named "getByShortLink" since now we have GET /shortLinks/:shortLink available?

**Comment by fm3** [comment]
right, let’s leave it there then :) yes, I still think getByShortLink is a good idea to mark that this is a special value

**Review by fm3 — COMMENTED** [review]
LGTM! Let’s wait for the frontend before merging

**Comment by philippotto** [comment]
> @philippotto maybe you have some time to build this in the coming days? Otherwise I will try my hand :)
> We also did not implement the actual redirect in the backend (since the backend usually only answers /api calls). So it would be cool if the frontend could use the new backend routes to perform the redirect when the user visits `/link/asdf`

I'll try to squeeze it in this week :)

**Comment by fm3** [comment]
Thanks!

By the way, I’m not super happy with one particular naming, but I don’t have a better idea. Maybe you have suggestions? The `ShortLink` case class (good name) has in turn a field `shortLink: String`. I find it a bit weird that this is the same name. Also the shortLink value does not start with http, but is only the generated, secret id/token/string that identifies the link. For the zarr private links, @leowe called a similar string `tokenValue`, which might also be an option here, but I’m not super happy with that either. Open for suggestions!

**Comment by frcroth** [comment]
> Thanks!
> 
> By the way, I’m not super happy with one particular naming, but I don’t have a better idea. Maybe you have suggestions? The `ShortLink` case class (good name) has in turn a field `shortLink: String`. I find it a bit weird that this is the same name. Also the shortLink value does not start with http, but is only the generated, secret id/token/string that identifies the link. For the zarr private links, @leowe called a similar string `tokenValue`, which might also be an option here, but I’m not super happy with that either, since we have so many other tokens already… Open for suggestions!

Both are not properly named with the term Shortlink. The case class actually stores a *link mapping*, while the short link does not store a link but the final part of a path ^^.

**Comment by philippotto** [comment]
> Also the shortLink value does not start with http, but is only the generated, secret id/token/string that identifies the link.

[wikipedia](https://en.wikipedia.org/wiki/URL_shortening) refers to this term as `key` which makes sense to me. I think, there exists another name for this term in URLs (similar to stump, thunk or something like that?), but I cannot remember it anymore. `key` would be fine for me.

> The case class rather stores some kind of link mapping

True, but I think `ShortLink` is still okay for the case class, as the purpose (admittedly not content) is described well :)

**Comment by philippotto** [comment]
The front-end is done now :)

**Review by fm3 — APPROVED** [review]
> wikipedia refers to this term as key

Good enough for me :) I took the liberty to change this in the code so that you don’t have to go over everything again. In the process I also read and tested the frontend code. Works for me :)

@frcroth Could you add an entry in the migration guide linking to the sql evolution? That’s https://github.com/scalableminds/webknossos/blob/master/MIGRATIONS.unreleased.md – and you can have a look at https://github.com/scalableminds/webknossos/blob/master/MIGRATIONS.released.md for inspiration (when we draft a monthly release, the entries from the unreleased changelog and migration guide are moved to released)

Then this should be good to go :shipit:  (talk to me when merging to master, as the evolution has to be manually applied on the production server)

**Pull request merged**
